boot/LoadReactOS.c

```diff
@@ -0,0 +1,473 @@
+	long dwSize;
+	OPTMULTIBOOT *multiboot = &config->opt.Multiboot;
+
+	memset(FREELDR_LOAD_AREA, 0, FREELDR_MAX_SIZE);
+
+	VIDEO_ATTR = 0xffd8d8d8;
+	printk("  Loading %s from CD", config->szPath);
+	VIDEO_ATTR = 0xffa8a8a8;
+	dwSize = BootIso9660GetFile(config->drive, config->szPath, FREELDR_LOAD_AREA, FREELDR_MAX_SIZE);
+
+	if (dwSize < 0) {
+		printk("Not Found, error %d\nHalting\n", dwSize); 
+		wait_ms(2000);
+		return false;
+	} else {
+		printk(" - %d bytes\n", dwSize);
+
+		if (CheckMultibootHeader(FREELDR_LOAD_AREA) == NULL) {
+			wait_ms(2000);
+			return false;
+		}
+		multiboot->pBuffer = FREELDR_LOAD_AREA;
+		multiboot->uBufferSize = dwSize;
+		/* Multiboot partition indices starting at 0, FreeLoader will increment 0xFE to 0xFF
+		 * and later it will interpret partition number 0xFF as CD drive.
+		 * DL should be set to 0x80 for HDD, and 0xE0 for CD-ROM. */
+		multiboot->uBootDevice = GetMultibootBootDevice(0xE0, 0xFE, 0xFF, 0xFF);
+	}
+
+	return true;
+}
+
+int ExittoReactOS(const OPTMULTIBOOT *multiboot) {
+	PMULTIBOOTHEADER mbHeader;
+
+	mbHeader = CheckMultibootHeader(multiboot->pBuffer);
+	if (mbHeader == NULL)
+		return;
+
+	VIDEO_ATTR = 0xff8888a8;
+	printk("     Multiboot header found at 0x%X\n", mbHeader);
+	printk("     Boot device is 0x%X\n", multiboot->uBootDevice);
+	VIDEO_ATTR = 0xffa8a8a8;
+
+	char *sz = "\2Starting ReactOS\2";
+	VIDEO_CURSOR_POSX = ((vmode.width - BootVideoGetStringTotalWidth(sz)) / 2) * 4;
+	VIDEO_CURSOR_POSY = vmode.height - 64;
+
+	VIDEO_ATTR = 0xff9f9fbf;
+	printk(sz);
+	setLED("rrrr");
+	startReactOS(mbHeader, multiboot->uBufferSize, multiboot->uBootDevice);
+}
+
+void startReactOS(PMULTIBOOTHEADER mbHeader, u32 loaderSize, u32 bootDevice) {
+	PMULTIBOOTINFO mbInfo;
+	PMEMORYMAP mmap;
+	static u32 ldrSize;
+	static u32 bootDev;
+	u32 infoPtr;
+	int nAta = 0;
+
+	if (mbHeader == NULL)
+		return;
+
+	/* Save needed values into static variables, as the stack pointer
+	 * will be modified by inline assembly below */
+	ldrSize = loaderSize;
+	bootDev = bootDevice;
+
+	/* Prepare to move freeldr to its final destination. Since that might
+	 * involve overwriting cromwell structures we shutdown as much as
+	 * possible. After this point, we can't return anymore */
+	BootStopUSB();
+	if (tsaHarddiskInfo[0].m_bCableConductors == 80) {
+		if (tsaHarddiskInfo[0].m_wAtaRevisionSupported & 2) nAta = 1;
+		if (tsaHarddiskInfo[0].m_wAtaRevisionSupported & 4) nAta = 2;
+		if (tsaHarddiskInfo[0].m_wAtaRevisionSupported & 8) nAta = 3;
+		if (tsaHarddiskInfo[0].m_wAtaRevisionSupported & 16) nAta = 4;
+		if (tsaHarddiskInfo[0].m_wAtaRevisionSupported & 32) nAta = 5;
+	} else {
+		// force the HDD into a good mode 0x40 ==UDMA | 2 == UDMA2
+		nAta = 2; // best transfer mode without 80-pin cable
+	}
+	// nAta=1;
+	BootIdeSetTransferMode(0, 0x40 | nAta);
+	BootIdeSetTransferMode(1, 0x40 | nAta);
+
+	/* orange, people seem to like that colour */
+	setLED("oooo");
+
+	/* disable interrupts */
+	asm volatile ("cli\n");
```

Comment:
Up until this point this, the function seems to be the same code as the Linux variant. We should eventually factor this out.